File: forge/crates/forge-io/src/zip.rs
//! A minimal, deterministic ZIP container (PKWARE APPNOTE 6.3, no ZIP64, no
//! encryption) for OPC packages such as 3MF.
//!
//! # Writer
//! - Entries are written **sorted by name** (byte order), each with a fixed timestamp
//!   (1980-01-01 00:00:00, the DOS epoch), no extra fields, no comments and no data
//!   descriptors, so the same entries always produce the same bytes.
//! - Data is compressed with raw DEFLATE ([`miniz_oxide`], level 6) and stored instead
//!   when that is not smaller.
//!
//! # Reader
//! Parses the end-of-central-directory record and the central directory, checks every
//! local header against it, inflates (method 8) or copies (method 0) the data with the
//! declared size as a hard limit, and verifies the CRC-32. Encrypted entries, ZIP64 and
//! other methods are rejected with [`IoError::Zip`].
//!
//! Only the DEFLATE codec is borrowed; the container logic and CRC-32 are ours.

use std::collections::BTreeMap;

use crate::IoError;

const LOCAL_SIG: u32 = 0x0403_4b50;
const CENTRAL_SIG: u32 = 0x0201_4b50;
const EOCD_SIG: u32 = 0x0605_4b50;
/// DOS date for 1980-01-01 (`(year − 1980) << 9 | month << 5 | day`).
const DOS_DATE_EPOCH: u16 = (1 << 5) | 1;
const DOS_TIME_MIDNIGHT: u16 = 0;
const VERSION_20: u16 = 20;
const METHOD_STORE: u16 = 0;
const METHOD_DEFLATE: u16 = 8;
const DEFLATE_LEVEL: u8 = 6;

/// CRC-32 (IEEE 802.3, reflected polynomial `0xEDB88320`) lookup table.
const CRC_TABLE: [u32; 256] = {
    let mut t = [0u32; 256];
    let mut i = 0;
    while i < 256 {
        let mut c = i as u32;
        let mut k = 0;
        while k < 8 {
            c = if c & 1 != 0 {
                0xEDB8_8320 ^ (c >> 1)
            } else {
                c >> 1
            };
            k += 1;
        }
        t[i] = c;
        i += 1;
    }
    t
};

/// CRC-32 of `data` (as used by ZIP, PNG, gzip).
pub fn crc32(data: &[u8]) -> u32 {
    let mut c = 0xFFFF_FFFFu32;
    for &b in data {
        c = CRC_TABLE[((c ^ u32::from(b)) & 0xFF) as usize] ^ (c >> 8);
    }
    c ^ 0xFFFF_FFFF
}

fn zip_err(detail: impl Into<String>) -> IoError {
    IoError::Zip {
        detail: detail.into(),
    }
}

fn put16(out: &mut Vec<u8>, v: u16) {
    out.extend_from_slice(&v.to_le_bytes());
}
fn put32(out: &mut Vec<u8>, v: u32) {
    out.extend_from_slice(&v.to_le_bytes());
}

fn to_u32(n: usize, what: &str) -> Result<u32, IoError> {
    u32::try_from(n).map_err(|_| zip_err(format!("{what} exceeds 4 GiB (ZIP64 not supported)")))
}

/// Write a ZIP archive with the given `(name, data)` entries (sorted by name).
pub fn write_zip(entries: &[(&str, &[u8])]) -> Result<Vec<u8>, IoError> {
    let mut sorted: Vec<&(&str, &[u8])> = entries.iter().collect();
    sorted.sort_by(|a, b| a.0.as_bytes().cmp(b.0.as_bytes()));
    for w in sorted.windows(2) {
        if w[0].0 == w[1].0 {
            return Err(zip_err(format!("duplicate entry {}", w[0].0)));
        }
    }
    if sorted.len() > usize::from(u16::MAX) {
        return Err(zip_err("too many entries"));
    }
    let mut out = Vec::new();
    let mut central = Vec::new();
    for (name, data) in sorted {
        if name.is_empty() || !name.is_ascii() || name.len() > usize::from(u16::MAX) {
            return Err(zip_err(format!("invalid entry name {name:?}")));
        }
        let deflated = miniz_oxide::deflate::compress_to_vec(data, DEFLATE_LEVEL);
        let (method, payload): (u16, &[u8]) = if deflated.len() < data.len() {
            (METHOD_DEFLATE, &deflated)
        } else {
            (METHOD_STORE, data)
        };
        let crc = crc32(data);
        let offset = to_u32(out.len(), "archive")?;
        let csize = to_u32(payload.len(), "entry")?;
        let usize_ = to_u32(data.len(), "entry")?;
        let name_len = name.len() as u16;
        // Local file header.
        put32(&mut out, LOCAL_SIG);
        put16(&mut out, VERSION_20);
        put16(&mut out, 0);
        put16(&mut out, method);
        put16(&mut out, DOS_TIME_MIDNIGHT);
        put16(&mut out, DOS_DATE_EPOCH);
        put32(&mut out, crc);
        put32(&mut out, csize);
        put32(&mut out, usize_);
        put16(&mut out, name_len);
        put16(&mut out, 0);
        out.extend_from_slice(name.as_bytes());
        out.extend_from_slice(payload);
        // Central directory header.
        put32(&mut central, CENTRAL_SIG);
        put16(&mut central, VERSION_20); // made by: MS-DOS, 2.0
        put16(&mut central, VERSION_20);
        put16(&mut central, 0);
        put16(&mut central, method);
        put16(&mut central, DOS_TIME_MIDNIGHT);
        put16(&mut central, DOS_DATE_EPOCH);
        put32(&mut central, crc);
        put32(&mut central, csize);
        put32(&mut central, usize_);
        put16(&mut central, name_len);
        put16(&mut central, 0); // extra
        put16(&mut central, 0); // comment
        put16(&mut central, 0); // disk
        put16(&mut central, 0); // internal attributes
        put32(&mut central, 0); // external attributes
        put32(&mut central, offset);
        central.extend_from_slice(name.as_bytes());
    }
    let cd_offset = to_u32(out.len(), "archive")?;
    let cd_size = to_u32(central.len(), "central directory")?;
    let count = u16::try_from(entries.len()).map_err(|_| zip_err("too many entries"))?;
    out.extend_from_slice(&central);
    put32(&mut out, EOCD_SIG);
    put16(&mut out, 0);
    put16(&mut out, 0);
    put16(&mut out, count);
    put16(&mut out, count);
    put32(&mut out, cd_size);
    put32(&mut out, cd_offset);
    put16(&mut out, 0);
    Ok(out)
}

struct Reader<'a> {
    b: &'a [u8],
}

impl Reader<'_> {
    fn u16(&self, at: usize) -> Result<u16, IoError> {
        self.b
            .get(at..at + 2)
            .map(|s| u16::from_le_bytes([s[0], s[1]]))
            .ok_or_else(|| zip_err("truncated archive"))
    }
    fn u32(&self, at: usize) -> Result<u32, IoError> {
        self.b
            .get(at..at + 4)
            .map(|s| u32::from_le_bytes([s[0], s[1], s[2], s[3]]))
            .ok_or_else(|| zip_err("truncated archive"))
    }
    fn slice(&self, at: usize, len: usize) -> Result<&[u8], IoError> {
        at.checked_add(len)
            .and_then(|end| self.b.get(at..end))
            .ok_or_else(|| zip_err("truncated archive"))
    }
}

/// Read every entry of a ZIP archive: name → uncompressed bytes (sorted by name).
pub fn read_zip(bytes: &[u8]) -> Result<BTreeMap<String, Vec<u8>>, IoError> {
    let r = Reader { b: bytes };
    if bytes.len() < 22 {
        return Err(zip_err("not a zip archive (too short)"));
    }
    // End of central directory: scan back over a possible comment (≤ 65535 bytes).
    let lowest = bytes.len().saturating_sub(22 + 65_535);
    let mut eocd = None;
    let mut at = bytes.len() - 22;
    loop {
        if r.u32(at)? == EOCD_SIG {
            eocd = Some(at);
            break;
        }
        if at == lowest {
            break;
        }
        at -= 1;
    }
    let eocd = eocd.ok_or_else(|| zip_err("end of central directory not found"))?;
    let count = usize::from(r.u16(eocd + 10)?);
    let cd_size = r.u32(eocd + 12)? as usize;
    let cd_offset = r.u32(eocd + 16)? as usize;
    if cd_offset == 0xFFFF_FFFF || count == 0xFFFF {
        return Err(zip_err("ZIP64 archives are not supported"));
    }
    r.slice(cd_offset, cd_size)?;
    let mut out = BTreeMap::new();
    let mut p = cd_offset;
    for _ in 0..count {
        if r.u32(p)? != CENTRAL_SIG {
            return Err(zip_err("bad central directory header"));
        }
        let flags = r.u16(p + 8)?;
        let method = r.u16(p + 10)?;
        let crc = r.u32(p + 16)?;
        let csize = r.u32(p + 20)? as usize;
        let usize_ = r.u32(p + 24)? as usize;
        let name_len = usize::from(r.u16(p + 28)?);
        let extra_len = usize::from(r.u16(p + 30)?);
        let comment_len = usize::from(r.u16(p + 32)?);
        let local = r.u32(p + 42)? as usize;
        let name_bytes = r.slice(p + 46, name_len)?;
        let name = String::from_utf8(name_bytes.to_vec())
            .map_err(|_| zip_err("entry name is not UTF-8"))?;
        if flags & 1 != 0 {
            return Err(zip_err(format!("entry {name} is encrypted")));
        }
        if csize == 0xFFFF_FFFF || usize_ == 0xFFFF_FFFF {
            return Err(zip_err("ZIP64 entries are not supported"));
        }
        if r.u32(local)? != LOCAL_SIG {
            return Err(zip_err(format!("bad local header for {name}")));
        }
        let lname = usize::from(r.u16(local + 26)?);
        let lextra = usize::from(r.u16(local + 28)?);
        let data = r.slice(local + 30 + lname + lextra, csize)?;
        let content = match method {
            METHOD_STORE => {
                if csize != usize_ {
                    return Err(zip_err(format!(
                        "stored entry {name} has inconsistent sizes"
                    )));
                }
                data.to_vec()
            }
            METHOD_DEFLATE => miniz_oxide::inflate::decompress_to_vec_with_limit(data, usize_)
                .map_err(|e| zip_err(format!("entry {name}: inflate failed ({:?})", e.status)))?,
            m => return Err(zip_err(format!("entry {name}: unsupported method {m}"))),
        };
        if content.len() != usize_ {
            return Err(zip_err(format!("entry {name}: size mismatch")));
        }
        if crc32(&content) != crc {
            return Err(zip_err(format!("entry {name}: CRC mismatch")));
        }
        if out.insert(name.clone(), content).is_some() {
            return Err(zip_err(format!("duplicate entry {name}")));
        }
        p += 46 + name_len + extra_len + comment_len;
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn crc32_matches_the_standard_check_value() {
        assert_eq!(crc32(b"123456789"), 0xCBF4_3926);
        assert_eq!(crc32(b""), 0);
    }

    #[test]
    fn zip_round_trips_sorted_and_deterministic() {
        let big = vec![b'a'; 10_000];
        let entries: [(&str, &[u8]); 3] = [("z.txt", b"zz"), ("a/b.bin", &big), ("m", b"")];
        let z1 = write_zip(&entries).expect("zip");
        let z2 = write_zip(&entries).expect("zip");
        assert_eq!(z1, z2);
        let back = read_zip(&z1).expect("read");
        let names: Vec<&str> = back.keys().map(String::as_str).collect();
        assert_eq!(names, ["a/b.bin", "m", "z.txt"]);
        assert_eq!(back["a/b.bin"], big);
        assert_eq!(back["z.txt"], b"zz");
        // Sorted in the file too: the first local header is a/b.bin.
        assert_eq!(&z1[30..37], b"a/b.bin");
        // Deflate kicked in for the compressible entry.
        assert!(z1.len() < 1_000);
    }

    #[test]
    fn corrupted_archives_are_rejected() {
        let z = write_zip(&[("x", b"hello hello hello hello")]).expect("zip");
        // Payload corruption: CRC mismatch or inflate failure.
        let mut bad = z.clone();
        bad[33] ^= 0x55;
        assert!(read_zip(&bad).is_err());
        // Broken local header signature.
        let mut bad = z.clone();
        bad[0] ^= 0xFF;
        assert!(read_zip(&bad).is_err());
        // Truncation.
        assert!(read_zip(&z[..z.len() - 1]).is_err());
        assert!(read_zip(b"not a zip at all, definitely not").is_err());
        assert!(write_zip(&[("a", b"1"), ("a", b"2")]).is_err());
    }
}

File: packages/app/src/file/partzero/zip.ts
/**
 * Our own zip container, host-free.
 *
 * {@link writeZip} is deterministic: entries sorted by name, a fixed timestamp (1980-01-01 00:00, the zip epoch), no
 * extra fields, no comments, no OS attributes, UTF-8 names, and our own DEFLATE ({@link deflateRaw}). The same entries
 * always give the same bytes, on every platform.
 *
 * {@link readZip} reads what we write and what ordinary zip tools write (stored and deflated entries, data
 * descriptors), and refuses what a document never needs and an attacker might use: encryption, ZIP64, multi-disk
 * archives, unsafe names (absolute, `..`, backslashes, drive letters, control characters), duplicate names, and
 * entries or totals above the limits (a zip bomb stops at the cap, see {@link inflateRaw}).
 */
import { crc32 } from "./checksum";
import { deflateRaw, inflateRaw, InflateError } from "./deflate";

export type ZipErrorCode = "ZIP_NOT_A_ZIP" | "ZIP_UNSUPPORTED" | "ZIP_CORRUPT" | "ZIP_UNSAFE_NAME" | "ZIP_TOO_LARGE" | "ZIP_DUPLICATE";

export class ZipError extends Error {
  readonly code: ZipErrorCode;
  readonly entry: string | null;
  constructor(code: ZipErrorCode, message: string, entry: string | null = null) {
    super(message);
    this.name = "ZipError";
    this.code = code;
    this.entry = entry;
  }
}

export interface ZipEntryInput {
  name: string;
  data: Uint8Array;
  /** Compress this entry (kept stored when compression does not make it smaller). Default true. */
  compress?: boolean;
}

export interface ZipLimits {
  /** Most entries (default 10 000). */
  maxEntries: number;
  /** Largest uncompressed entry (default 1 GiB). */
  maxEntryBytes: number;
  /** Largest uncompressed total (default 2 GiB). */
  maxTotalBytes: number;
}

export const DEFAULT_ZIP_LIMITS: ZipLimits = { maxEntries: 10_000, maxEntryBytes: 1024 ** 3, maxTotalBytes: 2 * 1024 ** 3 };

const SIG_LOCAL = 0x04034b50;
const SIG_CENTRAL = 0x02014b50;
const SIG_EOCD = 0x06054b50;
const FLAG_UTF8 = 0x0800;
/** 1980-01-01: day 1, month 1, year 0 (1980) in MS-DOS date format. */
const DOS_DATE = (0 << 9) | (1 << 5) | 1;
const DOS_TIME = 0;
const MAX_NAME = 512;

const utf8 = new TextEncoder();

/**
 * Compressed forms of large inputs, by identity: a document's imported meshes are the same `Uint8Array` objects from
 * one save (or autosave) to the next, so they are compressed once. The output is a pure function of the bytes, so the
 * cache never changes a file. (A caller that mutates an array after writing it must not rely on this cache: nothing in
 * the document layer does; blobs are immutable.)
 */
const deflated = new WeakMap<Uint8Array, { crc: number; body: Uint8Array | null }>();
const CACHE_FROM_BYTES = 64 * 1024;

function compressEntry(data: Uint8Array): { crc: number; body: Uint8Array | null } {
  const hit = data.length >= CACHE_FROM_BYTES ? deflated.get(data) : undefined;
  if (hit) return hit;
  const z = deflateRaw(data);
  const r = { crc: crc32(data), body: z.length < data.length ? z : null };
  if (data.length >= CACHE_FROM_BYTES) deflated.set(data, r);
  return r;
}

/** Whether `name` is a safe, relative, forward-slash entry name (the only kind we write or read). */
export function isSafeEntryName(name: string): boolean {
  if (name.length === 0 || name.length > MAX_NAME) return false;
  // eslint-disable-next-line no-control-regex
  if (/[\u0000-\u001f\u007f\\:*?"<>|]/.test(name)) return false;
  if (name.startsWith("/")) return false;
  const parts = name.replace(/\/$/, "").split("/");
  return parts.every((p) => p.length > 0 && p !== "." && p !== "..");
}

/** Build a zip archive from `entries` (deterministic; see the module comment). */
export function writeZip(entries: readonly ZipEntryInput[]): Uint8Array {
  const sorted = [...entries].sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  const seen = new Set<string>();
  const locals: Uint8Array[] = [];
  const centrals: Uint8Array[] = [];
  let offset = 0;
  for (const e of sorted) {
    if (!isSafeEntryName(e.name) || e.name.endsWith("/")) throw new ZipError("ZIP_UNSAFE_NAME", `invalid entry name: ${JSON.stringify(e.name)}`, e.name);
    if (seen.has(e.name)) throw new ZipError("ZIP_DUPLICATE", `duplicate entry: ${e.name}`, e.name);
    seen.add(e.name);
    const name = utf8.encode(e.name);
    let crc: number;
    let method = 0;
    let body = e.data;
    if (e.compress !== false && e.data.length > 0) {
      const c = compressEntry(e.data);
      crc = c.crc;
      if (c.body) {
        method = 8;
        body = c.body;
      }
    } else {
      crc = crc32(e.data);
    }
    if (body.length > 0xfffffffe || e.data.length > 0xfffffffe) throw new ZipError("ZIP_TOO_LARGE", `entry too large for a zip without ZIP64: ${e.name}`, e.name);
    const local = new Uint8Array(30 + name.length);
    const lv = new DataView(local.buffer);
    lv.setUint32(0, SIG_LOCAL, true);
    lv.setUint16(4, 20, true); // version needed: 2.0
    lv.setUint16(6, FLAG_UTF8, true);
    lv.setUint16(8, method, true);
    lv.setUint16(10, DOS_TIME, true);
    lv.setUint16(12, DOS_DATE, true);
    lv.setUint32(14, crc, true);
    lv.setUint32(18, body.length, true);
    lv.setUint32(22, e.data.length, true);
    lv.setUint16(26, name.length, true);
    lv.setUint16(28, 0, true);
    local.set(name, 30);
    locals.push(local, body);

    const central = new Uint8Array(46 + name.length);
    const cv = new DataView(central.buffer);
    cv.setUint32(0, SIG_CENTRAL, true);
    cv.setUint16(4, 20, true); // made by: MS-DOS, 2.0 (no Unix attributes)
    cv.setUint16(6, 20, true);
    cv.setUint16(8, FLAG_UTF8, true);
    cv.setUint16(10, method, true);
    cv.setUint16(12, DOS_TIME, true);
    cv.setUint16(14, DOS_DATE, true);
    cv.setUint32(16, crc, true);
    cv.setUint32(20, body.length, true);
    cv.setUint32(24, e.data.length, true);
    cv.setUint16(28, name.length, true);
    // extra, comment, disk, internal and external attributes: 0
    cv.setUint32(42, offset, true);
    central.set(name, 46);
    centrals.push(central);
    offset += local.length + body.length;
    if (offset > 0xfffffffe) throw new ZipError("ZIP_TOO_LARGE", "archive too large for a zip without ZIP64");
  }
  if (sorted.length > 0xfffe) throw new ZipError("ZIP_TOO_LARGE", "too many entries for a zip without ZIP64");
  const cdSize = centrals.reduce((s, c) => s + c.length, 0);
  const eocd = new Uint8Array(22);
  const ev = new DataView(eocd.buffer);
  ev.setUint32(0, SIG_EOCD, true);
  ev.setUint16(8, sorted.length, true);
  ev.setUint16(10, sorted.length, true);
  ev.setUint32(12, cdSize, true);
  ev.setUint32(16, offset, true);
  const out = new Uint8Array(offset + cdSize + eocd.length);
  let p = 0;
  for (const part of [...locals, ...centrals, eocd]) {
    out.set(part, p);
    p += part.length;
  }
  return out;
}

export interface ZipEntry {
  name: string;
  data: Uint8Array;
}

function findEocd(bytes: Uint8Array, view: DataView): number {
  // The EOCD record is 22 bytes plus a comment of up to 65535 bytes, at the very end.
  const min = Math.max(0, bytes.length - 22 - 0xffff);
  for (let i = bytes.length - 22; i >= min; i--) {
    if (view.getUint32(i, true) === SIG_EOCD && i + 22 + view.getUint16(i + 20, true) === bytes.length) return i;
  }
  return -1;
}

const utf8Decoder = new TextDecoder("utf-8", { fatal: true });
const latin1Decoder = new TextDecoder("latin1");

/** Read every file entry of a zip archive (directory entries are skipped), checking CRCs and the limits. */
export function readZip(bytes: Uint8Array, limits: Partial<ZipLimits> = {}): ZipEntry[] {
  const lim = { ...DEFAULT_ZIP_LIMITS, ...limits };
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (bytes.length < 22) throw new ZipError("ZIP_NOT_A_ZIP", "not a zip archive (too short)");
  const eocd = findEocd(bytes, view);
  if (eocd < 0) throw new ZipError("ZIP_NOT_A_ZIP", "not a zip archive (no end-of-central-directory record)");
  const disk = view.getUint16(eocd + 4, true);
  const cdDisk = view.getUint16(eocd + 6, true);
  const count = view.getUint16(eocd + 10, true);
  const cdSize = view.getUint32(eocd + 12, true);
  const cdOffset = view.getUint32(eocd + 16, true);
  if (disk !== 0 || cdDisk !== 0 || view.getUint16(eocd + 8, true) !== count) throw new ZipError("ZIP_UNSUPPORTED", "multi-disk zip archives are not supported");
  if (count === 0xffff || cdSize === 0xffffffff || cdOffset === 0xffffffff) throw new ZipError("ZIP_UNSUPPORTED", "ZIP64 archives are not supported");
  if (cdOffset + cdSize > eocd) throw new ZipError("ZIP_CORRUPT", "the central directory lies outside the archive");
  if (count > lim.maxEntries) throw new ZipError("ZIP_TOO_LARGE", `too many entries (${count} > ${lim.maxEntries})`);
  const out: ZipEntry[] = [];
  const names = new Set<string>();
  let total = 0;
  let p = cdOffset;
  for (let k = 0; k < count; k++) {
    if (p + 46 > cdOffset + cdSize || view.getUint32(p, true) !== SIG_CENTRAL) throw new ZipError("ZIP_CORRUPT", "damaged central directory");
    const flags = view.getUint16(p + 8, true);
    const method = view.getUint16(p + 10, true);
    const crc = view.getUint32(p + 16, true);
    const csize = view.getUint32(p + 20, true);
    const usize = view.getUint32(p + 24, true);
    const nameLen = view.getUint16(p + 28, true);
    const extraLen = view.getUint16(p + 30, true);
    const commentLen = view.getUint16(p + 32, true);
    const localOffset = view.getUint32(p + 42, true);
    const nameBytes = bytes.subarray(p + 46, p + 46 + nameLen);
    p += 46 + nameLen + extraLen + commentLen;
    let name: string;
    try {
      name = flags & FLAG_UTF8 ? utf8Decoder.decode(nameBytes) : latin1Decoder.decode(nameBytes);
    } catch {
      throw new ZipError("ZIP_UNSAFE_NAME", "an entry name is not valid UTF-8");
    }
    if (!isSafeEntryName(name)) throw new ZipError("ZIP_UNSAFE_NAME", `unsafe entry name: ${JSON.stringify(name.slice(0, 200))}`, name);
    if (flags & 0x0001) throw new ZipError("ZIP_UNSUPPORTED", `encrypted entries are not supported (${name})`, name);
    if (csize === 0xffffffff || usize === 0xffffffff || localOffset === 0xffffffff) throw new ZipError("ZIP_UNSUPPORTED", "ZIP64 entries are not supported", name);
    if (names.has(name)) throw new ZipError("ZIP_DUPLICATE", `duplicate entry: ${name}`, name);
    names.add(name);
    if (name.endsWith("/")) continue; // a directory
    if (method !== 0 && method !== 8) throw new ZipError("ZIP_UNSUPPORTED", `compression method ${method} is not supported (${name})`, name);
    if (usize > lim.maxEntryBytes) throw new ZipError("ZIP_TOO_LARGE", `${name} is too large (${usize} bytes)`, name);
    total += usize;
    if (total > lim.maxTotalBytes) throw new ZipError("ZIP_TOO_LARGE", "the archive's contents are too large");
    if (localOffset + 30 > bytes.length || view.getUint32(localOffset, true) !== SIG_LOCAL) throw new ZipError("ZIP_CORRUPT", `damaged local header (${name})`, name);
    const start = localOffset + 30 + view.getUint16(localOffset + 26, true) + view.getUint16(localOffset + 28, true);
    if (start + csize > bytes.length) throw new ZipError("ZIP_CORRUPT", `${name} is truncated`, name);
    const raw = bytes.subarray(start, start + csize);
    let data: Uint8Array;
    if (method === 0) {
      if (csize !== usize) throw new ZipError("ZIP_CORRUPT", `${name}: stored size mismatch`, name);
      data = raw.slice();
    } else {
      try {
        data = inflateRaw(raw, { expectedSize: usize, maxSize: Math.min(lim.maxEntryBytes, usize) });
      } catch (e) {
        if (e instanceof InflateError && e.code === "INFLATE_TOO_LARGE") throw new ZipError("ZIP_CORRUPT", `${name} inflates beyond its recorded size`, name);
        throw new ZipError("ZIP_CORRUPT", `${name}: ${e instanceof Error ? e.message : String(e)}`, name);
      }
    }
    if (data.length !== usize) throw new ZipError("ZIP_CORRUPT", `${name}: size mismatch`, name);
    if (crc32(data) !== crc) throw new ZipError("ZIP_CORRUPT", `${name}: checksum mismatch`, name);
    out.push({ name, data });
  }
  return out;
}

/** Whether `bytes` starts like a zip archive (`PK\x03\x04`, or an empty archive's `PK\x05\x06`). */
export function looksLikeZip(bytes: Uint8Array): boolean {
  return bytes.length >= 4 && bytes[0] === 0x50 && bytes[1] === 0x4b && ((bytes[2] === 3 && bytes[3] === 4) || (bytes[2] === 5 && bytes[3] === 6));
}
